File: 0019-remove-nth-node-from-end-of-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        //coutn no. of nodes
        ListNode temp=head;
        int count=0;
        while(temp != null){
            count++;
            temp=temp.next;
        }
        
        //if we have to delete 1st node i.e  count==n 
        if( count == n){
            head=head.next;
            return head;
        }
       // else 1st  reach at the 1 node before of target node 
       temp=head;
        int diff=count-n;
        for(int i=1; i<diff; i++){
            temp=temp.next;
        }
        //here we are at 1 node before of target(delete node) so 
        temp.next=temp.next.next;
        
        return head;
    }
}

File: 0019-remove-nth-node-from-end-of-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // 1st count the no of nodes 
        int cnt = 0;
        ListNode* temp = head;
        while(temp != NULL){
            cnt++;
            temp = temp->next;
        }

        int nFromStart = cnt - n + 1;
        if(nFromStart == 1) return head->next;

        temp = head;
        int tempCnt = 1;
        while( true ){
            if(tempCnt == nFromStart-1) break;
            tempCnt++;
            temp = temp->next;
        }

        temp->next = temp->next->next;
        return head;
    }
};
